Src/Fido2.Models/CredentialCreateOptions.cs

```diff
@@ -61,7 +61,7 @@ public sealed class CredentialCreateOptions : Fido2ResponseBase
     /// This member is intended for use by Relying Parties that wish to limit the creation of multiple credentials for the same account on a single authenticator.The client is requested to return an error if the new credential would be created on an authenticator that also contains one of the credentials enumerated in this parameter.
     /// </summary>
     [JsonPropertyName("excludeCredentials")]
-    public List<PublicKeyCredentialDescriptor> ExcludeCredentials { get; set; }
+    public IEnumerable<PublicKeyCredentialDescriptor> ExcludeCredentials { get; set; }

```

Comment:
There's supposed to be actual data here - how does a serializer know what to deserialise here? Do POCOs/DTOs really need interfaces on their data?

Reply:
Why not use IReadOnlyList<T> or IList<T> and still have an efficient .Count property and indexer if the goal is a common interface for lists and arrays?

I'd only use IEnumerable<T> to explicitly support lazy loading / streaming processing.

Reply:
Sure, the library does not do anything with it, it just passes it through. But with that rationale one might as well advocate for using `object` ;)
The API consumer should also be considered, and they are being helped by having an abstraction that closely matches that data in the DTO. The data here will always be an array-style-thing and `IList` allows the API consumer to use their favorite flavor of that. `IEnumerable` is a magic generator box with potentially unbounded output, and requires the consumer to query element by element just to get a count. Even though we already know the actual object being transported has to be able to provide that information at near-zero cost.

Also from the perspective of minimizing surprises `IEnumerable` is questionable in this case. You remove `.ToList()` here:
```
var existingKeys = _demoStorage.GetCredentialsByUser(user).Select(c => c.Descriptor).ToList();
var options = _fido2.RequestNewCredential(user, existingKeys, ...);
return Json(options);
```
In a typical setup, the storage will be something like EntityFramework, and now `existingKeys` is a prepared SQL query. If there happens to be a problem with the database, the exception will be thrown in the bowels of the Json serializer, when the prepared query will finally be executed on the DB. And often controller functions don't even explicitly serialize, but returns the `options` objects directly leave that to the ASP.NET middleware, making for an even more interesting stack trace.

Reply:
> and requires the consumer to query element by element just to get a count.

The consumer, the library here, doesn't need to know the count or alter the data.

@abergs, @aseigler any thought?

Reply:
The `CredentialCreateOptions` DTO is public API.
What would you call Blazor, if not an API consumer?

Reply:
Not sure to understand why you want to change it to something more specific than `IEnumerable<PublicKeyCredentialDescriptor>`.

The fido lib itself is not doing more than returning it,

https://github.com/passwordless-lib/fido2-net-lib/blob/eb312de4c580fb43c28f8e9928d0dcd6c1fec76d/Src/Fido2.Models/CredentialCreateOptions.cs#L99

and the value which will be serialised, or not, later.

It's the same thing for `AllowCredentials` in `AssertionOptions`

https://github.com/passwordless-lib/fido2-net-lib/blob/eb312de4c580fb43c28f8e9928d0dcd6c1fec76d/Src/Fido2.Models/AssertionOptions.cs#L39-L39

which is already a `IEnumerable`.

Reply:
> Not sure to understand why you want to change it to something more specific
> The fido lib itself is not doing more than returning it, and the value which will be serialised, or not, later.

And a Blazor app may be deserializing it (into the same class) on the client side, and now has the inconvenience of `IEnumerable<T>`, instead of an array-like type. Which is why you had to change Blazor code for this PR.

And suggesting to users in the readme to pass in not-yet evaluated EF queries, which may only get executed later after the context has already been disposed is setting up another tripwire.

> It's the same thing for `AllowCredentials` in `AssertionOptions` which is already a `IEnumerable`

Which could be improved by changing that to `IList<T>` as suggested in #191

Reply:
I agree if I prefer IList (or anything collection-like) rather than IEnumerable, since it's definitely completed the moment I receive or pass it somewhere.

Reply:
Changed to `IReadOnlyList`. Demo app no longer impacted by this change 👍